src/GSoft.Extensions.Http.Authentication.ClientCredentialsGrant/OpenIdConfigurationRetrieverWithCache.cs

```diff
@@ -0,0 +1,60 @@
+﻿using System.Collections.Concurrent;
+using System.Text.Json;
+
+namespace GSoft.Extensions.Http.Authentication.ClientCredentialsGrant;
+
+/// <summary>
+/// Class responsible for retrieving the OpenID metadata configuration from an identity provider.
+/// Inspired from Microsoft.Identity.Client's new generic authority feature:
+/// https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/4.53.0/src/client/Microsoft.Identity.Client/Instance/Oidc/OidcRetrieverWithCache.cs
+/// </summary>
+internal sealed class OpenIdConfigurationRetrieverWithCache : IOpenIdConfigurationRetriever, IDisposable
+{
+    private readonly IHttpClientFactory _httpClientFactory;
+    private readonly ConcurrentDictionary<string, OpenIdConfiguration> _cache;
+    private readonly SemaphoreSlim _mutex;
+
+    public OpenIdConfigurationRetrieverWithCache(IHttpClientFactory httpClientFactory)
+    {
+        this._httpClientFactory = httpClientFactory;
+        this._cache = new ConcurrentDictionary<string, OpenIdConfiguration>(StringComparer.OrdinalIgnoreCase);
+        this._mutex = new SemaphoreSlim(1);
+    }
+
+    public async Task<OpenIdConfiguration> GetAsync(string authority, CancellationToken cancellationToken)
+    {
+        if (this._cache.TryGetValue(authority, out var configuration))
+        {
+            return configuration;
+        }
+
+        await this._mutex.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+        try
+        {
+            if (this._cache.TryGetValue(authority, out configuration))
+            {
+                return configuration;
+            }
+
+            var httpClient = this._httpClientFactory.CreateClient(ClientCredentialsConstants.BackchannelHttpClientName);
+            var oidcMetadataEndpoint = authority.TrimEnd('/') + "/.well-known/openid-configuration";
```

Comment:
Would it be better to build the Url and path using a Uri object so that way trailing `/` and other formatting are taken care of?

Reply:
Yes, that's possible. I haven't done it in the first place because:

* Microsoft doesn't do it in MSAL ([ref](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/4.53.0/src/client/Microsoft.Identity.Client/Instance/Oidc/OidcRetrieverWithCache.cs#L42)) - and actually they don't even take care of trimming a potential trailing `/`
* `authority` has already been validated as a well-formed absolute URI

I agree that using `Uri` is generally safer when it comes to combining and handling URIs, especially in web apps for `redirectUrl` handling. But you must know what you're doing.

Combining URIs with the `Uri` class can be error-prone too. For instance, consider this example:

```csharp
new Uri(new Uri("https://authority/foo/bar"), "/.well-known/openid-configuration")
```

One might expect `https://authority/foo/bar/.well-known/openid-configuration` as a result. But NOPE. The actual result is `https://authority/.well-known/openid-configuration`. Having a leading `/` in the relative part wipes out any existing path.

Also consider this last example:

```csharp
new Uri(new Uri("https://authority/foo/bar"), ".well-known/openid-configuration")
```

Without the leading slash on the relative part, you might think that you're good. But you're not. This is the actual result: `https://authority/foo/.well-known/openid-configuration`. Notice how `bar` got removed.

For all those reasons, and because of my two first points, I'd like to keep the code as-is.